tensorflow-framework/src/main/java/org/tensorflow/framework/activations/ReLU.java

```diff
@@ -0,0 +1,143 @@
+ *     result = relu.call(input);
+ *     // result is [0.f, 0.f, 0.f, 5.f, 5.f]
+ *
+ *     // With threshold = 5
+ *     relu = new ReLU&lt;&gt;(tf, ReLU.ALPHA_DEFAULT, ReLU.MAX_VALUE_DEFAULT, 5f);
+ *     result = relu.call(input);
+ *     // result is [-0.f, -0.f,  0.f,  0.f, 10.f]
+ * </pre>
+ *
+ * @param <T> the data type of the result
+ */
+public class ReLU<T extends TNumber> extends Activation<T> {
+
+  public static final float ALPHA_DEFAULT = 0.0f;
+  public static final float MAX_VALUE_DEFAULT = Float.NaN;
+  public static final float THRESHOLD_DEFAULT = 0.0f;
+
+  private final float alpha;
+  private final float maxValue;
+  private final float threshold;
+
+  /**
+   * Creates a new ReLU with alpha={@link #ALPHA_DEFAULT}, maxValue={@link #MAX_VALUE_DEFAULT},
+   * threshold={@link #THRESHOLD_DEFAULT},
+   *
+   * @param tf the TensorFlow Ops
+   */
+  public ReLU(Ops tf) {
+    this(tf, ALPHA_DEFAULT, MAX_VALUE_DEFAULT, THRESHOLD_DEFAULT);
+  }
+
+  /**
+   * Creates a new ReLU
+   *
+   * @param tf the TensorFlow Ops
+   * @param alpha governs the slope for values lower than the threshold.
+   * @param maxValue sets the saturation threshold (the largest value the function will return).
+   * @param threshold the threshold value of the activation function below which values will be
+   *     damped or set to zero.
+   */
+  public ReLU(Ops tf, float alpha, float maxValue, float threshold) {
+    super(tf);
+    this.alpha = alpha;
+    this.maxValue = maxValue;
+    this.threshold = threshold;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Operand<T> call(Operand<T> input) {
+
+    DataType<T> dataType = input.asOutput().dataType();
+    Operand<T> negativePart = null;
+    boolean clipMax = !Float.isNaN(maxValue);
+    if (alpha != 0) {
+      if (Float.isNaN(maxValue) && threshold == 0) {
+        // TODO LeakyRelu is not in tf.nn ????
+        return LeakyRelu.create(tf.scope(), input, LeakyRelu.alpha(alpha));
+      }
+      if (threshold != 0) {
+        negativePart =
+            tf.nn.relu(
+                tf.math.add(tf.math.neg(input), tf.dtypes.cast(tf.constant(threshold), dataType)));
+      } else {
+        negativePart = tf.nn.relu(tf.math.neg(input));
+      }
+    }
+
+    if (threshold != 0) {
+      // computes input for input > threshold else 0
+      Greater greater = tf.math.greater(input, tf.dtypes.cast(tf.constant(threshold), dataType));
+      input = tf.math.mul(input, tf.dtypes.cast(greater, dataType));
+    } else if (maxValue == 6) {
+      // if no threshold, then can use nn.relu6 native TF op for performance
+      input = tf.nn.relu6(input);
+      clipMax = false;
+    } else {
+      input = tf.nn.relu(input);
+    }
+    if (clipMax) {
+      Operand<T> lmaxValue = tf.dtypes.cast(tf.constant(maxValue), dataType);
+      Operand<T> zero = tf.dtypes.cast(tf.constant(0), dataType);
+      input = tf.clipByValue(input, zero, lmaxValue);
```

Comment:
Is zero still the right lower clip if there's a threshold?

Reply:
The numbers less than the `threshold` get converted to `zero` before the `clipMax` is applied.
So, there is no need to move the lower bound on the `clipByValue` operation.

```
if (threshold != 0) {
      // computes input for input > threshold else 0
      Greater greater = tf.math.greater(input, tf.dtypes.cast(tf.constant(threshold), dataType));
      input = tf.math.mul(input, tf.dtypes.cast(greater, dataType));
```